mlir/lib/Dialect/SCF/IR/SCF.cpp

```diff
@@ -622,6 +626,47 @@ LogicalResult scf::ForallOp::promoteIfSingleIteration(RewriterBase &rewriter) {
   return success();
 }
 
+Block::BlockArgListType ForallOp::getRegionIterArgs() {
+  return getBody()->getArguments().drop_front(getRank());
+}
+
+MutableArrayRef<OpOperand> ForallOp::getInitsMutable() {
+  return getOutputsMutable();
+}
+
+FailureOr<LoopLikeOpInterface>
+ForallOp::replaceWithAdditionalYields(RewriterBase &rewriter,
+                                      ValueRange newInitOperands,
+                                      bool replaceInitOperandUsesInLoop,
+                                      const NewYieldValuesFn &newYieldValueFn) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(getOperation());
+  auto inits = llvm::to_vector(getOutputs());
+  inits.append(newInitOperands.begin(), newInitOperands.end());
+  auto newLoop = rewriter.create<scf::ForallOp>(
+      getLoc(), getMixedLowerBound(), getMixedUpperBound(), getMixedStep(),
+      inits, getMapping(), [](OpBuilder &, Location, ValueRange) {});
+
+  // Move the region of the current block to the newly created op.
+  Block *newLoopBody = newLoop.getBody();
+  rewriter.mergeBlocks(
+      getBody(), newLoopBody,
+      newLoopBody->getArguments().take_front(getBody()->getNumArguments()));
+
+  // Update the terminator.
+  {
+    OpBuilder::InsertionGuard g(rewriter);
+    auto terminator = cast<scf::InParallelOp>(newLoopBody->getTerminator());
+    rewriter.setInsertionPointToEnd(terminator.getBody());
+    newYieldValueFn(
```

Comment:
I couldn't find the place in the code, but I think this lambda creates the `parallel_insert_slice` ops. It probably returns an empty `SmallVector`. This does not fit with the comment in `LoopLikeInterface.h`:

```c++
/// A function that returns the additional yielded values during
/// `replaceWithAdditionalYields`. `newBbArgs` are the newly added region
/// iter_args. This function should return as many values as there are block
/// arguments in `newBbArgs`.
using NewYieldValuesFn = std::function<SmallVector<Value>(
    OpBuilder &b, Location loc, ArrayRef<BlockArgument> newBbArgs)>;
```

It is a bit confusing that for `scf.for` we expect the function to return a list of `Value` and for `scf.forall` we expect the function to build new IR and not return anything.

How about we generalize `NewYieldValuesFn` to `RewriteTerminatorFn = std::function<void(RewriterBase &rewriter, Operation *terminator, ArrayRef<BlockArgument> newBbArgs)>`?

Reply:
Yeah, but the `scf.forall` does not return any values. I looked at place where `NewYieldValuesFn` is used and it is quite a bit. Want to limit the breaking changes here, but I can do a quick follow up on that one.

Reply:
Actually that might not work as well. Consider https://github.com/llvm/llvm-project/blob/5ca2d75f2046612978ba71c4b36714b2a0a01886/mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp#L193 . This is written in a loop agnostic way, but yield value function will have to handle `scf.forall` separately. I think there is an inherent mismatch between the behavior of `scf.forall` and other loop constructs in the sense it doesnt yield anything.

Reply:
Actually I do have an issue here (and I didnt see it cause I am also missing a test that will show up this issue). So good catch... Ill drop this part and do something different...

Reply:
> I think there is an inherent mismatch between the behavior of `scf.forall` and other loop constructs in the sense it doesnt yield anything.

I ran into the same problem when cleaning up the `LoopLikeOpInterface` recently. I was able to add the interface to `scf.while`, but not `scf.forall` because of this issue.

I think we should generalize `NewYieldValuesFn` to `RewriteTerminatorFn`. That lambda can do whatever it wants.

```c++
/// A function that updates the terminator of a loop-like op during
/// `replaceWithAdditionalYields`. `newBbArgs` are the newly added region
/// iter_args. This function should rewrite the loop terminator (`terminator`)
/// such that the loop op verifies successfully.
using RewriteTerminatorFn =
    std::function<void(RewriterBase &rewriter, Operation *terminator, ArrayRef<BlockArgument> newBbArgs)>;
```

I think we also have to turn this interface function into an interface method:
```c++
    /// Append the specified additional "init" operands: replace this loop with
    /// a new loop that has the additional init operands. The loop body of this
    /// loop is moved over to the new loop.
    ///
    /// The newly added region iter_args are yielded from the loop.
    ::mlir::FailureOr<::mlir::LoopLikeOpInterface>
        replaceWithAdditionalIterOperands(::mlir::RewriterBase &rewriter,
                                          ::mlir::ValueRange newInitOperands,
                                          bool replaceInitOperandUsesInLoop)
```

For `scf.forall`, this interface could either:
- "return failure();"
- or: insert a parallel_insert_slice that inserts the newly added bbArg into itself (which is a no-op)
- or: remove this interface method entirely

Reply:
The implementations of `yieldTiledValuesAndReplace` make sense to me. But I wouldn't put it in `LoopLikeOpInterface`.

`LoopLikeOpInterface` was meant to expose basic loop properties and low-level transformations of loop-like ops. The new `yieldTiledValuesAndReplace` interface method is tiling-specific and does not belong in `LoopLikeOpInterface` imo. How about putting `yieldTiledValuesAndReplace` in a new interface in `TilingInterface.td`? Then we have all tiling-related stuff in one place.

Reply:
I am not sure a whole interface for just one method is worth it. I think this method is optional for the most part. No op needs to implement it. It is needed only if it wants to work with tiling, which is nice cause all the loop construct has to do to work with tiling is implement this method. Id prefer to keep it in `LoopLikeOpInterface`. If you feel strongly about it, I can move it out of the interface.